firedrake/mg/generalised_hierarchy.py

```diff
@@ -0,0 +1,236 @@
+""" Generates Mesh and FunctionSpace hierarchies that can have different
+refinement factors of multiples of 2 """
+
+from __future__ import absolute_import
+
+from firedrake import *  # noqa
+from firedrake.mg import *  # noqa
+from firedrake.mg.utils import *  # noqa
+
+
+class GeneralisedMeshHierarchy(object):
+
+    """ Builds a MeshHierarchy with a user defined refinement factor (multiple of 2)
+
+        :param mesh: The coarsest mesh
+        :type mesh: :class:`Mesh'
+
+        :param levels: number of levels (multiple of M)
+        :type levels: int
+
+        :param M: Refinement factor (multiple of 2)
+        :type M: int
+
+    """
+
+    def __init__(self, mesh, levels, M=2):
+
+        # refinement factor
+        self.M = M
+
+        # check that M is a multiple of 2
+        if (M % 2) != 0:
+            raise ValueError('Refinement factor is not a multiple of 2')
+
+        # 'skip parameter'
+        self.skip = (M / 2)
+
+        # carry the full mesh hierarchy
+        self._full_hierarchy = MeshHierarchy(mesh, levels * self.skip)
+
+        # carry the generalised mesh hierarchy
+        self._hierarchy = self._full_hierarchy[::self.skip]
+
+        super(GeneralisedMeshHierarchy, self).__init__()
```

Comment:
I'm pretty sure that you don't need a new set of classes for this at all.  I think if you just do the following in `MeshHierarchy`:

```python
def __init__(self, mesh, levels, refinements_per_level=1):
   ...
   hierarchy = all_the_refinements
   self._full_hierarchy = hierarchy
   self._hierarchy = hierarchy[::refinements_per_level]
   self.refinements_per_level = refinements_per_level
```

Then everything just falls out.  Similarly in functionspacehierarchy and friends.